src/common/python.ts

```diff
@@ -17,87 +67,66 @@ interface IExtensionApi {
-        onDidChangePythonInterpreterEvent.fire({
-            path: execCommand,
-            resource,
-        });
-    }
-}
-
-async function getPythonExtensionAPI(): Promise<IExtensionApi | undefined> {
+async function activateExtension() {
     const extension = extensions.getExtension('ms-python.python');
     if (extension) {
         if (!extension.isActive) {
             await extension.activate();
         }
     }
+    return extension;
+}
 
+async function getPythonExtensionAPI(): Promise<IExtensionApi | undefined> {
+    const extension = await activateExtension();
     return extension?.exports as IExtensionApi;
 }
 
 export async function initializePython(disposables: Disposable[]): Promise<void> {
     try {
-        interpreterMap.set('', undefined);
-        getWorkspaceFolders().forEach((w) => interpreterMap.set(w.uri.toString(), undefined));
-
-        const extension = extensions.getExtension('ms-python.python');
-        if (extension) {
-            if (!extension.isActive) {
-                await extension.activate();
-            }
-
-            const api = await getPythonExtensionAPI();
-
-            if (api) {
-                disposables.push(
-                    api.settings.onDidChangeExecutionDetails((resource) => {
-                        updateInterpreterFromExtension(api, resource);
-                    }),
-                );
-
-                updateInterpreterFromExtension(api);
-            }
+        const api = await getPythonExtensionAPI();
+
+        if (api) {
+            disposables.push(
+                api.environment.onDidActiveEnvironmentChanged((e) => {
+                    onDidChangePythonInterpreterEvent.fire({ path: [e.path], resource: e.resource });
+                }),
+            );
+
+            traceLog('Waiting for interpreter from python extension.');
+            onDidChangePythonInterpreterEvent.fire(await getInterpreterDetails());
         }
     } catch (error) {
         traceError('Error initializing python: ', error);
     }
 }
 
 export async function getInterpreterDetails(resource?: Uri): Promise<IInterpreterDetails> {
-    const workspaceFolder = resource ? getWorkspaceFolder(resource) : undefined;
-
-    const key = workspaceFolder?.uri.toString() ?? '';
-    const interpreterPath = interpreterMap.get(key);
-    if (interpreterPath) {
-        return { path: interpreterPath, resource };
-    }
-
     const api = await getPythonExtensionAPI();
-    if (api) {
-        const execCommand = api.settings.getExecutionDetails(resource).execCommand;
-        interpreterMap.set(key, execCommand);
-        return { path: execCommand, resource };
+    const interpreterPath = await api?.environment.getActiveEnvironmentPath(resource);
+    if (interpreterPath) {
+        const details = await api?.environment.getEnvironmentDetails(interpreterPath.path);
+        if (details) {
+            return { path: [details.interpreterPath], resource };
+        }
     }
```

Comment:
```suggestion
    if (interpreterPath) {
        return { path: [interpreterPath.path], resource };
    }
```
This would work for all cases except when server is started using conda environments without python, which is an anomaly as we install python into those environments anyway.

Reply:
@karthiknadig If `interpreterPath.pathType` is `envFolderPath`, do not return it. I have edited the above suggestion to reflect that.